README.md
[![NPM Version](https://img.shields.io/npm/v/push-docker-image.svg?style=flat)](https://www.npmjs.org/package/push-docker-image)
[![Build Status](https://img.shields.io/travis/holidaycheck/push-docker-image/master.svg?style=flat)](https://travis-ci.org/holidaycheck/push-docker-image)
[![Coverage Status](https://img.shields.io/coveralls/holidaycheck/push-docker-image/master.svg?style=flat)](https://coveralls.io/r/holidaycheck/push-docker-image)
[![NPM Downloads](https://img.shields.io/npm/dm/push-docker-image.svg?style=flat)](https://www.npmjs.org/package/push-docker-image)
# push-docker-image

Push docker image `.tar.gz` files to a docker registry.

This tool is only compatible with:

* images following the [docker image spec v1.2](https://github.com/moby/moby/blob/5072b22c5fea93f00917c8c5d6a29d782db2bb73/image/spec/v1.2.md#combined-image-json--filesystem-changeset-format)
* registries that support the [registry API v2](https://docs.docker.com/registry/spec/api/)

This tool accepts a path to a docker image stored as a `.tar.gz` file following the [`Combined Image JSON + Filesystem Changeset Format` of the docker image specification v1.2](https://github.com/moby/moby/blob/5072b22c5fea93f00917c8c5d6a29d782db2bb73/image/spec/v1.2.md#combined-image-json--filesystem-changeset-format).

## Usage

### CLI

```
push-docker-image myImage.tar.gz
```

### Node API

```js
const pushDockerImage = require('push-docker-image');

const options = {
  auth: { username: 'user', password: 'password '},
  ssl: false
};

// The `options` argument is optional
pushDockerImage('/path/to/myImage.tar.gz', options)
    .then(() => console.log('Successfully uploaded.')
    .catch(() => console.log('Upload failed');
```


CHANGELOG.md
## 1.1.0 (May 31, 2019)

### Features

* Add `options` arguments for basic auth and ssl ([#45](https://github.com/holidaycheck/push-docker-image/pull/45))

### Dependency Upgrades

* Update dependency node-fetch to v2.6.0 ([#44](https://github.com/holidaycheck/push-docker-image/pull/44))
* Update dependency sinon to v7.3.2 ([#39](https://github.com/holidaycheck/push-docker-image/pull/39))
* Update dependency tmp to v0.1.0 ([#37](https://github.com/holidaycheck/push-docker-image/pull/37))
* Update dependency nyc to v14 ([#42](https://github.com/holidaycheck/push-docker-image/pull/42))
* Update dependency pretty-bytes to v5.2.0 ([#43](https://github.com/holidaycheck/push-docker-image/pull/43))
* Update dependency ava to v1.4.1 ([#38](https://github.com/holidaycheck/push-docker-image/pull/38))
* Update dependency tar-fs to v2 ([#28](https://github.com/holidaycheck/push-docker-image/pull/28))
* Update dependency hasha to v5 ([#41](https://github.com/holidaycheck/push-docker-image/pull/41))
* Update dependency eslint to v5.16.0 ([#40](https://github.com/holidaycheck/push-docker-image/pull/40))
* Update dependency coveralls to v3.0.3 ([#31](https://github.com/holidaycheck/push-docker-image/pull/31))
* Update dependency eslint to v5.15.3 ([#35](https://github.com/holidaycheck/push-docker-image/pull/35))
* Update dependency rimraf to v2.6.3 ([#26](https://github.com/holidaycheck/push-docker-image/pull/26))
* Update dependency sinon to v7.3.0 ([#30](https://github.com/holidaycheck/push-docker-image/pull/30))
* Update dependency hasha to v4.0.1 ([#34](https://github.com/holidaycheck/push-docker-image/pull/34))
* Update dependency ava to v1.4.0 ([#36](https://github.com/holidaycheck/push-docker-image/pull/36))
* Remove vulnerabilities in few packages ([#33](https://github.com/holidaycheck/push-docker-image/pull/33))
* Update dependency hasha to v4 ([#32](https://github.com/holidaycheck/push-docker-image/pull/32))
* Update dependency sinon to v7.2.3 ([#24](https://github.com/holidaycheck/push-docker-image/pull/24))
* Update dependency pr-log to v3.1.0 ([#27](https://github.com/holidaycheck/push-docker-image/pull/27))
* Update dependency ramda to v0.26.1 ([#22](https://github.com/holidaycheck/push-docker-image/pull/22))
* Update dependency ramda to v0.26.0 ([#21](https://github.com/holidaycheck/push-docker-image/pull/21))
* Update dependency node-fetch to v2.3.0 ([#20](https://github.com/holidaycheck/push-docker-image/pull/20))
* Update dependency eslint to v5.9.0 ([#19](https://github.com/holidaycheck/push-docker-image/pull/19))
* Update dependency node-fetch to v2.2.1 ([#18](https://github.com/holidaycheck/push-docker-image/pull/18))
* Update dependency sinon to v7.1.1 ([#17](https://github.com/holidaycheck/push-docker-image/pull/17))
* Update dependency eslint to v5.8.0 ([#16](https://github.com/holidaycheck/push-docker-image/pull/16))
* Update dependency sinon to v7.1.0 ([#15](https://github.com/holidaycheck/push-docker-image/pull/15))
* Update dependency noop3 to v1000 ([#13](https://github.com/holidaycheck/push-docker-image/pull/13))
* Update dependency nyc to v13 ([#12](https://github.com/holidaycheck/push-docker-image/pull/12))
* Update dependency node-fetch to v2.2.0 ([#11](https://github.com/holidaycheck/push-docker-image/pull/11))
* Update dependency eslint to v5.7.0 ([#10](https://github.com/holidaycheck/push-docker-image/pull/10))
* Update dependency sinon to v7 ([#14](https://github.com/holidaycheck/push-docker-image/pull/14))
* Update dependency tar-fs to v1.16.3 ([#6](https://github.com/holidaycheck/push-docker-image/pull/6))
* Update dependency sinon to v6 ([#8](https://github.com/holidaycheck/push-docker-image/pull/8))
* Update dependency eslint to v5 ([#7](https://github.com/holidaycheck/push-docker-image/pull/7))

### Build-Related

* Add missing coveralls dependency ([#4](https://github.com/holidaycheck/push-docker-image/pull/4))

## 1.0.0 (July 6, 2018)

Initial release


test/unit/lib/registryClientSpec.js
'use strict';

const test = require('ava');
const sinon = require('sinon');
const noop = require('noop3');

const { Readable } = require('stream');
const { Headers } = require('node-fetch');
const createRegistryClient = require('../../../lib/registryClient');

const imageDetails = {
    registry: 'example.com',
    repository: 'any-repo'
};

test('checks for existing blob', async (t) => {
    const fetch = sinon.stub().resolves({ status: 200 });

    const client = createRegistryClient(imageDetails, fetch);
    const readStream = new Readable();

    await client.uploadBlob(readStream, 'anyDigest').catch(noop);

    t.true(fetch.calledOnce);

    const [ url, options ] = fetch.firstCall.args;

    t.is(url, 'http://example.com/v2/any-repo/blobs/anyDigest');
    t.deepEqual(options, { method: 'HEAD' });
});

test('doesn’t upload anything and returns size from the content-length header when blob exists', async (t) => {
    const headers = new Headers({ 'Content-Length': 42 });
    const fetch = sinon.stub().resolves({ status: 200, headers });

    const client = createRegistryClient(imageDetails, fetch);
    const readStream = new Readable();

    const result = await client.uploadBlob(readStream, 'anyDigest').catch(noop);

    t.is(fetch.callCount, 1);
    t.deepEqual(result, { size: 42, digest: 'anyDigest' });
});

test('initiates a chunked blob upload correctly', async (t) => {
    const fetch = sinon.stub().resolves({ status: 418 });

    const client = createRegistryClient(imageDetails, fetch);
    const readStream = new Readable();

    await client.uploadBlob(readStream, 'anyDigest').catch(noop);

    t.is(fetch.callCount, 2);

    const [ url, options ] = fetch.secondCall.args;

    t.is(url, 'http://example.com/v2/any-repo/blobs/uploads/');
    t.deepEqual(options, { method: 'POST' });
});

test('rejects when the response status for initiating the upload is not 202', async (t) => {
    const fetch = sinon.stub().resolves({ status: 418 });

    const client = createRegistryClient(imageDetails, fetch);
    const readStream = new Readable();

    const error = await t.throwsAsync(() => client.uploadBlob(readStream, 'anyDigest'), Error);
    t.is(error.message, 'Failed to initiate blob upload to http://example.com.');
});

test('uses the URL from the Location header of the initiate response for subsequent requests', async (t) => {
    const fetch = sinon.stub();

    fetch.onFirstCall().resolves({ status: 404 });

    const responseHeaders = new Headers({ Location: 'http://example.com/foo/bar' });
    fetch.onSecondCall().resolves({ status: 202, headers: responseHeaders });

    fetch.onThirdCall().resolves({ status: 418 });

    const client = createRegistryClient(imageDetails, fetch);
    const readStream = new Readable();

    readStream.push('foo');
    readStream.push(null);

    await client.uploadBlob(readStream, 'anyDigest').catch(noop);

    t.is(fetch.callCount, 3);

    const [ url ] = fetch.thirdCall.args;
    t.is(url, 'http://example.com/foo/bar');
});

test('finalizes an upload after all chunks haven been uploaded', async (t) => {
    const fetch = sinon.stub();

    fetch.onFirstCall().resolves({ status: 404 });

    const responseHeaders = new Headers({ Location: '/foo/bar' });
    fetch.onSecondCall().resolves({ status: 202, headers: responseHeaders });

    const uploadResponseHeaders = new Headers({ Location: 'http://example.com/foo/bar?a=b' });
    fetch.onThirdCall().resolves({ status: 202, headers: uploadResponseHeaders });
    fetch.onCall(3).resolves({ status: 418 });

    const client = createRegistryClient(imageDetails, fetch);
    const readStream = new Readable();

    readStream.push('foo');
    readStream.push(null);

    await client.uploadBlob(readStream, 'anyDigest').catch(noop);

    t.is(fetch.callCount, 4);

    const [ fourthCallUrl, fourthCallOptions ] = fetch.lastCall.args;

    const expectedDigiest = 'anyDigest';
    t.is(fourthCallUrl, `http://example.com/foo/bar?a=b&digest=${encodeURIComponent(expectedDigiest)}`);

    t.deepEqual(fourthCallOptions, { method: 'PUT', headers: { 'Content-Length': 0 } });
});

test('rejects an error when the finialize response status is not 201', async (t) => {
    const fetch = sinon.stub();

    const responseHeaders = new Headers({ Location: '/foo/bar' });
    fetch.onFirstCall().resolves({ status: 404 });
    fetch.onSecondCall().resolves({ status: 202, headers: responseHeaders });

    fetch.onThirdCall().resolves({ status: 202, headers: responseHeaders });
    fetch.onCall(3).resolves({ status: 418 });

    const client = createRegistryClient(imageDetails, fetch);
    const readStream = new Readable();

    readStream.push('foo');
    readStream.push(null);

    const error = await t.throwsAsync(() => client.uploadBlob(readStream, 'anyDigest'), Error);

    t.is(error.message, 'Failed to complete blob upload to http://example.com.');
});

test('does a complete upload and resolves with the digest and size', async (t) => {
    const fetch = sinon.stub();

    const responseHeaders = new Headers({ Location: 'http://example.com/foo/bar' });
    fetch.onFirstCall().resolves({ status: 404 });
    fetch.onSecondCall().resolves({ status: 202, headers: responseHeaders });

    fetch.onThirdCall().resolves({ status: 202, headers: responseHeaders });
    fetch.onCall(3).resolves({ status: 201, headers: responseHeaders });

    const client = createRegistryClient(imageDetails, fetch);
    const readStream = new Readable();

    readStream.push('foo');
    readStream.push(null);

    const result = await client.uploadBlob(readStream, 'anyDigest');

    t.deepEqual(result, {
        digest: 'anyDigest',
        size: 3
    });
});

test('calls the onProgress callback for each uploaded chunk and on completion', async (t) => {
    const fetch = sinon.stub();
    const onProgress = sinon.stub();

    fetch.onFirstCall().resolves({ status: 404 });

    const responseHeaders = new Headers({ Location: 'http://example.com/foo/bar' });
    fetch.onSecondCall().resolves({ status: 202, headers: responseHeaders });

    fetch.onThirdCall().resolves({ status: 202, headers: responseHeaders });
    fetch.onCall(3).resolves({ status: 202, headers: responseHeaders });
    fetch.onCall(4).resolves({ status: 201, headers: responseHeaders });

    const client = createRegistryClient(imageDetails, fetch);
    const readStream = new Readable();

    readStream.push('foo');
    readStream.push('bar');
    readStream.push(null);

    await client.uploadBlob(readStream, 'anyDigest', { onProgress });

    t.is(onProgress.callCount, 3);
    t.deepEqual(onProgress.firstCall.args, [ { size: 3, done: false, alreadyExists: false } ]);
    t.deepEqual(onProgress.secondCall.args, [ { size: 6, done: false, alreadyExists: false } ]);
    t.deepEqual(onProgress.thirdCall.args, [ { size: 6, done: true, alreadyExists: false } ]);
});

test('calls the onProgress callback immediately when blob already exists', async (t) => {
    const headers = new Headers({ 'Content-Length': 42 });
    const fetch = sinon.stub().resolves({ status: 200, headers });

    const client = createRegistryClient(imageDetails, fetch);
    const readStream = new Readable();

    const onProgress = sinon.stub();
    await client.uploadBlob(readStream, 'anyDigest', { onProgress }).catch(noop);

    t.is(onProgress.callCount, 1);
    t.deepEqual(onProgress.firstCall.args, [ { size: 42, done: true, alreadyExists: true } ]);
});

test('calls the onProgress callback immediately when blob already exists with SSL', async (t) => {
    const headers = new Headers({ 'Content-Length': 42 });
    const fetch = sinon.stub().resolves({ status: 200, headers });

    const client = createRegistryClient(imageDetails, fetch, { ssl: true });
    const readStream = new Readable();

    const onProgress = sinon.stub();
    await client.uploadBlob(readStream, 'anyDigest', { onProgress }).catch(noop);

    t.is(onProgress.callCount, 1);
    t.deepEqual(onProgress.firstCall.args, [ { size: 42, done: true, alreadyExists: true } ]);
});


test/unit/lib/createCustomFetchSpec.js
'use strict';

const test = require('ava');
const http = require('http');
const listen = require('test-listen');
const createCustomFetch = require('../../../lib/createCustomFetch');

async function customFetchMacro(t, { customFetchOptions, headers, expectedResult }) {
    const server = http.createServer((req, res) => {
        res.end(req.headers.authorization);
    });

    const url = await listen(server);

    const fetch = createCustomFetch(customFetchOptions);
    const response = await fetch(url, headers ? { headers } : undefined);
    const result = Boolean(await response.text());

    server.close();

    t.is(result, expectedResult);
}

test('has auth', customFetchMacro, {
    customFetchOptions: { username: 'test', password: 'test' },
    expectedResult: true
});

test('has no auth', customFetchMacro, {
    expectedResult: false
});

test('has auth with header', customFetchMacro, {
    customFetchOptions: { username: 'test', password: 'test' },
    headers: { 'x-test': 1 },
    expectedResult: true
});

test('has no auth with header', customFetchMacro, {
    headers: { 'x-test': 1 },
    expectedResult: false
});


test/unit/lib/extractImageDetailsSpec.js
'use strict';

const test = require('ava');
const extractImageDetails = require('../../../lib/extractImageDetails');

test('throws an error when the manifest contains more than one repository', (t) => {
    const manifest = [ { RepoTags: [ 1, 2 ] } ];
    const expectedErrorMessage = 'Unexpected amount of repositories. Got 2 but expected 1.';

    const error = t.throws(() => extractImageDetails(manifest), Error);
    t.is(error.message, expectedErrorMessage);
});

test('throws an error when the manifest contains less than one repository', (t) => {
    const manifest = [ { RepoTags: [] } ];
    const expectedErrorMessage = 'Unexpected amount of repositories. Got 0 but expected 1.';

    const error = t.throws(() => extractImageDetails(manifest), Error);
    t.is(error.message, expectedErrorMessage);
});

test('throws an error when the image name doesn’t contain a docker registry host', (t) => {
    const manifest = [ { RepoTags: [ 'foo:bar' ] } ];
    const expectedErrorMessage = 'No registry defined in image name "foo:bar".';

    const error = t.throws(() => extractImageDetails(manifest), Error);
    t.is(error.message, expectedErrorMessage);
});

test('returns all relevant image details', (t) => {
    const manifest = [ {
        RepoTags: [ 'example.com/foo:bar' ],
        Layers: [
            'layer-1/layer.tar',
            'layer-2/layer.tar'
        ],
        Config: 'config.json',
        something: 'not relevant'
    } ];
    const expectedDetails = {
        layers: [
            'layer-1/layer.tar',
            'layer-2/layer.tar'
        ],
        configFile: 'config.json',
        registry: 'example.com',
        repository: 'foo',
        tag: 'bar'
    };

    t.deepEqual(extractImageDetails(manifest), expectedDetails);
});

test('includes the image namespace in the repository name', (t) => {
    const manifest = [ {
        RepoTags: [ 'example.com/myNamespace/foo:bar' ],
        Layers: [ 'layer-1/layer.tar' ],
        Config: 'config.json'
    } ];
    const expectedDetails = {
        layers: [ 'layer-1/layer.tar' ],
        configFile: 'config.json',
        registry: 'example.com',
        repository: 'myNamespace/foo',
        tag: 'bar'
    };

    t.deepEqual(extractImageDetails(manifest), expectedDetails);
});


lib/pushDockerImage.js
'use strict';

const fs = require('fs');
const { extract } = require('tar-fs');
const gunzip = require('gunzip-maybe');
const path = require('path');
const pipeStreamsToPromise = require('pipe-streams-to-promise');
const tmp = require('tmp');
const { promisify } = require('util');
const rimraf = require('rimraf');
const hasha = require('hasha');
const extractImageDetails = require('./extractImageDetails');
const createRegistryClient = require('./registryClient');
const createCustomFetch = require('./createCustomFetch');
const formatBytes = require('pretty-bytes');
const zlib = require('zlib');

const deleteDir = promisify(rimraf);
const createTempDir = promisify(tmp.dir);

/* eslint-disable no-console */

async function extractArchive(dockerImageTarFile, dest) {
    const readArchiveStream = fs.createReadStream(dockerImageTarFile);
    const extractStream = extract(dest, { dmode: 0o666 });

    return pipeStreamsToPromise([ readArchiveStream, gunzip(), extractStream ]);
}

const chunkSize = 16 * 1024 * 1024;
const readStreamOptions = { highWaterMark: chunkSize };

async function compressLayer(layerPath) {
    if (layerPath.endsWith('.gz')) {
        return layerPath;
    }

    const readStream = fs.createReadStream(layerPath, readStreamOptions);
    const writeSteam = fs.createWriteStream(`${layerPath}.gz`, { encoding: 'binary' });
    const transformStream = zlib.createGzip({ chunkSize });

    await pipeStreamsToPromise([ readStream, transformStream, writeSteam ]);

    return `${layerPath}.gz`;
}

function createReporter(layerId) {
    return function ({ size, done, alreadyExists }) {
        if (!done) {
            console.log(`Layer ${layerId}: uploaded ${formatBytes(size)}.`);
        } else if (alreadyExists) {
            console.log(`Layer ${layerId}: already exists on registry.`);
        } else {
            console.log(`Layer ${layerId}: upload complete.`);
        }
    };
}

async function upload(tempDir, { auth, ssl }) {
    const imageManifest = require(path.join(tempDir, './manifest.json'));
    const imageDetails = extractImageDetails(imageManifest);

    const customFetch = createCustomFetch(auth);
    const registryClient = createRegistryClient(imageDetails, customFetch, { ssl });

    // eslint-disable-next-line max-statements
    const uploadAllLayerPromises = imageDetails.layers.map(async (relativeLayerPath) => {
        const [ layerId ] = relativeLayerPath.split('/');
        const layerPath = path.join(tempDir, relativeLayerPath);
        const compressedLayerPath = await compressLayer(layerPath);

        const hash = await hasha.fromFile(compressedLayerPath, { algorithm: 'sha256' });
        const digest = `sha256:${hash}`;
        const readStream = fs.createReadStream(compressedLayerPath, readStreamOptions);
        const options = { onProgress: createReporter(layerId) };

        return registryClient.uploadBlob(readStream, digest, options);
    });

    const imageConfigFile = path.join(tempDir, imageDetails.configFile);

    const imageConfigFileHash = await hasha.fromFile(imageConfigFile, { algorithm: 'sha256' });
    const [ imageConfigResult, ...layerResults ] = await Promise.all([
        registryClient.uploadBlob(fs.createReadStream(imageConfigFile), `sha256:${imageConfigFileHash}`),
        ...uploadAllLayerPromises
    ]);

    await registryClient.createImageDistributionManifest(layerResults, imageConfigResult);
}

module.exports = async function main(dockerImageTarFile, { auth, ssl } = {}) {
    const tempDir = await createTempDir();

    try {
        await extractArchive(dockerImageTarFile, tempDir);

        await upload(tempDir, { auth, ssl });

        // eslint-disable-next-line no-console
        console.log('Image successfully uploaded.');
    } finally {
        await deleteDir(tempDir);
    }
};


lib/createBlobWriteStream.js
'use strict';

const { Writable } = require('stream');

async function uploadBlobChunk(endpoint, lastRangeEnd, chunk, fetch) {
    const rangeStart = lastRangeEnd + 1;
    const rangeEnd = rangeStart + chunk.length - 1;
    const headers = {
        'Content-Length': chunk.length,
        'Content-Range': `${rangeStart}-${rangeEnd}`,
        'Content-Type': 'application/octet-stream'
    };

    const response = await fetch(endpoint, { method: 'PATCH', headers, body: chunk });

    if (response.status !== 202) {
        throw new Error('Failed to upload blob chunk.');
    }

    return response.headers.get('Location');
}

module.exports = function createBlobWriteStream(initialUploadUrl, fetch) {
    class BlobWriteStream extends Writable {
        constructor() {
            super({ highWaterMark: 16 * 1024 * 1024 });

            this.lastWrittenIndex = -1;
            this.uploadUrl = initialUploadUrl;
        }

        async _write(chunk, encoding, callback) {
            try {
                this.uploadUrl = await uploadBlobChunk(this.uploadUrl, this.lastWrittenIndex, chunk, fetch);
                this.lastWrittenIndex += chunk.length;
                this.emit('progress', this.lastWrittenIndex + 1);
                callback(null);
            } catch (error) {
                callback(error);
            }
        }
    }

    return new BlobWriteStream();
};


lib/registryClient.js
'use strict';

const noop = require('noop3');
const url = require('url');
const R = require('ramda');
const pipeStreamsToPromise = require('pipe-streams-to-promise');

const createBlobWriteStream = require('./createBlobWriteStream');
const buildDistributionManifest = require('./buildDistributionManifest');

module.exports = function createRegistryClient(imageDetails, fetch, { ssl } = {}) {
    const protocol = ssl ? 'https' : 'http';
    const registryUrl = `${protocol}://${imageDetails.registry}`;
    const { repository, tag } = imageDetails;

    async function initiateBlobUpload() {
        const startUploadEndpoint = `${registryUrl}/v2/${repository}/blobs/uploads/`;

        const response = await fetch(startUploadEndpoint, { method: 'POST' });

        if (response.status !== 202) {
            throw new Error(`Failed to initiate blob upload to ${registryUrl}.`);
        }

        return response.headers.get('Location');
    }

    async function finalizeChunkedUpload(endpoint, digest) {
        const headers = { 'Content-Length': 0 };
        const parsedEndpointUrl = R.dissoc('search', url.parse(endpoint, true));
        const endpointWithDigest = url.format(R.assocPath([ 'query', 'digest' ], digest, parsedEndpointUrl));

        const response = await fetch(endpointWithDigest, { method: 'PUT', headers });

        if (response.status !== 201) {
            throw new Error(`Failed to complete blob upload to ${registryUrl}.`);
        }
    }

    async function fetchExistingBlobDetails(digest) {
        const endpoint = `${registryUrl}/v2/${repository}/blobs/${digest}`;

        const response = await fetch(endpoint, { method: 'HEAD' });

        if (response.status === 200) {
            return {
                size: parseInt(response.headers.get('content-length'), 10)
            };
        }

        return null;
    }

    return {
        // eslint-disable-next-line max-statements
        async uploadBlob(readStream, digest, { onProgress = noop } = {}) {
            const existingBlob = await fetchExistingBlobDetails(digest);

            if (existingBlob) {
                onProgress({ size: existingBlob.size, done: true, alreadyExists: true });
                return { digest, size: existingBlob.size };
            }

            const uploadBlobEndpoint = await initiateBlobUpload();
            const writeStream = createBlobWriteStream(uploadBlobEndpoint, fetch);
            let size = 0;

            writeStream.on('progress', (currentSize) => {
                size = currentSize;
                onProgress({ size, done: false, alreadyExists: false });
            });

            await pipeStreamsToPromise([ readStream, writeStream ]);

            await finalizeChunkedUpload(writeStream.uploadUrl, digest);

            onProgress({ size, done: true, alreadyExists: false });

            return { digest, size };
        },

        async createImageDistributionManifest(layerResults, imageConfigResult) {
            const imageDistibutionManifest = buildDistributionManifest(layerResults, imageConfigResult);
            const headers = { 'Content-Type': imageDistibutionManifest.mediaType };
            const options = { method: 'PUT', headers, body: JSON.stringify(imageDistibutionManifest) };
            const endpoint = `${registryUrl}/v2/${repository}/manifests/${tag}`;

            const response = await fetch(endpoint, options);

            if (response.status !== 201) {
                throw new Error('Failed to push image distribution manifest.');
            }
        }
    };
};
